File: app/src/main/java/com/bitcoin/merchant/app/util/AddressUtil.java
package com.bitcoin.merchant.app.util;

import com.github.kiulian.converter.AddressConverter;
import com.github.kiulian.converter.b58.B58;

import de.tobibrandt.bitcoincash.BitcoinCashAddressFormatter;

public class AddressUtil {

    public static boolean isValidCashAddr(String address) {
        return BitcoinCashAddressFormatter.isValidCashAddress(address);
    }

    public static boolean isValidLegacy(String address) {
        try {
            B58.decodeAndCheck(address);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public static String toCashAddress(String legacy) {
        return AddressConverter.toCashAddress(legacy);
    }
}

File: app/src/main/java/com/bitcoin/merchant/app/util/AddressUtil.kt
package com.bitcoin.merchant.app.util

import org.bitcoinj.core.Address
import org.bitcoinj.core.AddressFormatException
import org.bitcoinj.core.CashAddressFactory
import org.bitcoinj.core.SlpAddress
import org.bitcoinj.params.MainNetParams

object AddressUtil {
    fun isValidCashAddr(cashAddress: String): Boolean {
        return try {
            Address.fromCashAddr(MainNetParams.get(), cashAddress)
            true
        } catch (e: AddressFormatException) {
            false
        }
    }

    fun isValidLegacy(legacyAddress: String): Boolean {
        return try {
            Address.fromBase58(MainNetParams.get(), legacyAddress)
            true
        } catch (e: AddressFormatException) {
            false
        }
    }

    fun toCashAddress(address: String): String {
        return if(isValidLegacy(address)) {
            CashAddressFactory.create().getFromBase58(MainNetParams.get(), address).toString()
        } else {
            CashAddressFactory.create().getFromFormattedAddress(MainNetParams.get(), address).toString()
        }
    }

    fun toLegacyAddress(address: String): String {
        return if(isValidCashAddr(address)) {
            CashAddressFactory.create().getFromFormattedAddress(MainNetParams.get(), address).toBase58()
        } else {
            CashAddressFactory.create().getFromBase58(MainNetParams.get(), address).toBase58()
        }
    }

    fun toSimpleLedgerAddress(address: String): String {
        return SlpAddress.fromCashAddr(MainNetParams.get(), address).toString()
    }

    fun fromSimpleLedgerAddress(address: String): String {
        return SlpAddress(MainNetParams.get(), address).toCashAddress()
    }
}